### DISCUSSION TITLE
RFC: Documentation overhaul

### DISCUSSION THREAD

**Discussion by LaserEyess** [transmission/transmission#3770] [Ideas]
This is a larger RFC with a few requests. Currently transmission's documentation lives in `docs/` as a series of markdown files that are only hosted on github, despite the fact that transmission has an actual website. Furthermore, API documentation is inline in `transmission.h` and not actually in any man pages or hosted on transmission's site. I think both of those things are bad, the first place people will go to look for documentation is the website, not the repository (though, the website does indeed link to the repository).

So, I would propose the following:
 - Use something like sphinx to build the markdown docs into HTML and link them directly on transmissionbt.com
 - Use doxygen+sphinx to turn `transmission.h` into HTML docs and host them on transmissionbt.com (this may require two sets of docs for the latest stable and the development API)
 - Keep these up to date by rebuilding the website every time a commit changes the documentation

Finally, this is a personal gripe and not directly relevant but still documentation related and would naturally follow from the above so I'd like to bring it up, RPC documentation should have a machine readable version so that other projects can ingest it and generate RPC data structures from it. This could also be used by transmission to build the RPC documentation. 

I'm willing to make some PRs for this since I'm familiar with sphinx, but I'd like feedback on whether devs are interested in this or if people have other suggestions.

**Comment by ckerr** [comment]
My main concerns:

- The documentation source needs to stay somewhere that contributors can make PRs for it. For example, it would be bad to go back to the GitHub wiki days where PRs aren't possible. The proposal here sounds like it's in line with this, but I wanted to point it out.

- Ideally we could update the website documentation automatically, e.g. I could hook some webhook bot up to make PRs for `transmission/transmission.github.io` when the generated docs have a diff from the ones in that repo. So ideally the product would be something that could be dumped & diff'ed relatively easily :smile_cat:

**Reply by LaserEyess** [reply]
>    * The documentation source needs to stay somewhere that contributors can make PRs for it. 

Correct, this would be the idea. One place to update the docs, one place that everyone can contribute to. 

> Ideally we could update the website documentation automatically, e.g. I could hook some webhook bot up to make PRs for `transmission/transmission.github.io` when the generated docs have a diff from the ones in that repo.

This is what [mpv currently does ](https://github.com/mpv-player/mpv.io) though their exact implementation may not work for transmission. Still, it is possible, and [even supported by github](https://github.com/marketplace/actions/github-pages-action) to some extent at least.

**Comment by LaserEyess** [comment]
@ckerr I'm going to take the thumbs up reaction as implicit support and start researching this. I assume adding python (via sphinx) as a build-only dep is acceptable?

**Reply by ckerr** [reply]
Yep sounds fine; thanks for asking :+1:
